### DISCUSSION TITLE
How to achieve equivariance for features of higher-order tensors?

### DISCUSSION THREAD

**Discussion by jiachenguoNU** [mir-group/nequip#274] [Q&A] [answered]
Hi,

I'm reading the Nature Communication paper and checking the code for NequIP and e3nn. In the paper, it is pointed out that the features on the atom can be tensors of different orders such as scalars, vectors and higher-order tensors. I can see from the example that scalars and vectors can be used as features (I guess that's because spherical harmonics are equivariant and we can project a vector function into the spherical harmonics.) 

But I don't know if I can use some other higher-order tensors such as the 2nd order symmetric tensor  (stress/strain) as the feature on the atoms. Could you please give some insights on this issue?

**Comment by Linux-cpp-lisp** [comment, marked as answer]
Yes, once you've decomposed whatever geometric input you have into a direct sum of irreducible representations, i.e. the e3nn "data format" (see the corresponding thread in e3nn https://github.com/e3nn/e3nn/discussions/376#discussioncomment-4277759), NequIP can take any such input as features on the atoms or edges.

You can see in the code where currently edge displacement vectors are projected onto the spherical harmonics for the edges; this module can be replaced; and similarly for the module that (in the default model) sets the node features to a scalar one-hot encoding of atom type.
